### PR TITLE
[XLA] Sink constants into the conditional computation in while loops

### PR DESCRIPTION
**Pull request by georgepaw**
Modify the WhileLoopConstantSinking  pass so that it sinks the constants into the while loop conditional as well. From my (limited) experiments, this helps the `WhileLoopAnalysis::ComputeWhileLoopTripCount` function in being able to evaluate and get the trip count (at least for simple while loops created by `dynamic_rnn`s where the original conditional is `and`ed with the max_iterations) .

### CODE REVIEW

File: tensorflow/compiler/xla/service/while_loop_constant_sinking.cc [hunk 1]
```diff
@@ -57,24 +58,44 @@ StatusOr<bool> WhileLoopConstantSinking::TrySinkingConstantsIntoWhileBody(
 
   bool changed = false;
 
-  for (HloInstruction* invariant_gte :
-       WhileUtil::GetInvariantGTEsForWhileBody(*while_body)) {
-    int64 index = invariant_gte->tuple_index();
+  auto invariant_conditional_gte_index_to_inst =
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/while_loop_constant_sinking.cc]: Let's use an explicit type for these.

File: tensorflow/compiler/xla/service/while_loop_constant_sinking.cc [hunk 2]
```diff
@@ -57,24 +58,44 @@ StatusOr<bool> WhileLoopConstantSinking::TrySinkingConstantsIntoWhileBody(
...
+  for (HloInstruction* invariant_body_gte : invariant_body_gtes) {
+    int64 index = invariant_body_gte->tuple_index();
     const HloInstruction& invariant_value = *init_value.operand(index);
 
-    // Should have at least one user that's not while_body_root.
-    if (invariant_gte->user_count() <= 1) {
-      continue;
-    }
+    // Original value should be a constant
+    if (invariant_value.opcode() != HloOpcode::kConstant) continue;
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/while_loop_constant_sinking.cc]: We always use braces for `if` and `for`.

File: tensorflow/compiler/xla/service/while_loop_constant_sinking.cc [hunk 3]
```diff
@@ -57,24 +58,44 @@ StatusOr<bool> WhileLoopConstantSinking::TrySinkingConstantsIntoWhileBody(
...
       auto* constant_instr =
           while_body->AddInstruction(invariant_value.Clone(/*suffix=*/".sunk"));
       TF_RETURN_IF_ERROR(ReplaceUsesWhileKeepingLoopInvariance(
-          invariant_gte, constant_instr, while_body->root_instruction(),
+          invariant_body_gte, constant_instr, while_body->root_instruction(),
           index));
       changed = true;
     }
+
+    // Check if there is a corresponding GTE in while_conditional
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/while_loop_constant_sinking.cc]: Please end comments with period (here and elsewhere).

File: tensorflow/compiler/xla/service/while_loop_constant_sinking.cc [hunk 4]
```diff
@@ -57,24 +58,44 @@ StatusOr<bool> WhileLoopConstantSinking::TrySinkingConstantsIntoWhileBody(
...
+    auto it = invariant_conditional_gte_index_to_inst.find(index);
+    if (it == invariant_conditional_gte_index_to_inst.end()) {
+      continue;
+    }
+
+    auto* invariant_cond_gte = it->second;
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/while_loop_constant_sinking.cc]: Would be nice to use an explicit type here and below.

File: tensorflow/compiler/xla/service/while_util.h [hunk 5]
```diff
@@ -84,6 +84,13 @@ class WhileUtil {
   // Assumes `while_body` is the body computation of the while loop in question.
   static std::vector<HloInstruction*> GetInvariantGTEsForWhileBody(
       const HloComputation& while_body);
+
+  // Returns a map of index to GetTupleElement instructions in
+  // `while_conditional` that access elements in the parameter tuple. Assumes
+  // `while_conditional` is the conditional computation of the while loop in
+  // question.
+  static std::map<int64, HloInstruction*> GetGTEsMapForWhileConditional(
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/while_util.h]: Do we care about stable ordering here?  If not, let's use ` absl::flat_hash_map`.

File: tensorflow/compiler/xla/service/while_util.cc [hunk 6]
```diff
@@ -268,4 +268,17 @@ static Shape MakeLoopStateShape(const WhileUtil::LoopStateTy& init_values) {
   return result;
 }
 
+/*static*/ std::map<int64, HloInstruction*>
+WhileUtil::GetGTEsMapForWhileConditional(
+    const HloComputation& while_conditional) {
+  std::map<int64, HloInstruction*> result;
+  for (auto* inst : while_conditional.instructions()) {
+    if (inst->opcode() == HloOpcode::kGetTupleElement &&
+        inst->operand(0) == while_conditional.parameter_instruction(0)) {
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/while_util.cc]: Can we instead walk the use list of `while_conditional.parameter_instruction(0)`?

File: tensorflow/compiler/xla/service/while_util.cc [hunk 7]
```diff
@@ -268,4 +268,17 @@ static Shape MakeLoopStateShape(const WhileUtil::LoopStateTy& init_values) {
...
+      result[inst->tuple_index()] = inst;
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/while_util.cc]: What if there are multiple GTEs with the same index?

**Review by sanjoy — CHANGES_REQUESTED** [review]

**Comment by georgepaw** [comment]
Thank you for the review comments! I think I have addressed all of them now.

**Review by sanjoy — CHANGES_REQUESTED** [review]
Just one minor nit.

File: tensorflow/compiler/xla/service/while_loop_constant_sinking.cc [hunk 8]
```diff
@@ -57,24 +58,48 @@ StatusOr<bool> WhileLoopConstantSinking::TrySinkingConstantsIntoWhileBody(
 
   bool changed = false;
 
-  for (HloInstruction* invariant_gte :
-       WhileUtil::GetInvariantGTEsForWhileBody(*while_body)) {
-    int64 index = invariant_gte->tuple_index();
+  absl::flat_hash_map<int64, absl::InlinedVector<HloInstruction*, 1>>
+      invariant_conditional_gte_index_to_inst =
```
[the remark this answers is not in the document]
Comment (georgepaw) [reply tensorflow/compiler/xla/service/while_loop_constant_sinking.cc]: Done!

**Comment by georgepaw** [comment]
Done, thank you 👍

**Review by sanjoy — APPROVED** [review]
Looks great!

**Pull request merged**